### PR TITLE
WIP fix existence hook issues.

### PR DESCRIPTION
**Pull request by heartpunk**
Failing test case adapted from comments in #1029; will either be squashed in once we see it passing or left as a failing commit as y'all prefer; for now it's just showing my work and setting up a relatively basic, clean starting spot to work from.

### CODE REVIEW

**Comment by heartpunk** [comment]
Ok, I'm going to take a break for a bit. Seems the last little bit of coverage (somewhat distant from what I refactored) is required, and I don't have a clear idea how to approach getting coverage there. It'll be clearer with time I'm sure.

File: packages/actor-query-operation-leftjoin/package.json [hunk 1]
```diff
@@ -38,6 +38,12 @@
     "sparqlalgebrajs": "^4.0.0",
     "sparqlee": "^2.1.0"
   },
+  "optionalDependencies": {
```
Comment (rubensworks) [inline packages/actor-query-operation-leftjoin/package.json:41]: These should not be necessary.

File: packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts [hunk 2]
```diff
@@ -1,16 +1,21 @@
...
 import { ArrayIterator, UnionIterator } from 'asynciterator';
+import * as chai from 'chai';
+import * as n3 from 'n3';
 import { DataFactory } from 'rdf-data-factory';
 import * as sparqlee from 'sparqlee';
 import { isExpressionError } from 'sparqlee';
-import { ActorQueryOperationLeftJoin } from '../lib/ActorQueryOperationLeftJoin';
+import { ActorQueryOperationLeftJoin } from '../lib';
 import '@comunica/jest';
+import 'mocha';
```
Comment (rubensworks) [inline packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts:14]: We're using jest in this project, so mocha is not required.

File: packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts [hunk 3]
```diff
@@ -1,16 +1,21 @@
...
 
 const DF = new DataFactory();
 const BF = new BindingsFactory();
+const sparql = new QueryEngine();
```
Comment (rubensworks) [inline packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts:18]: Since this is a unit test specific to this actor, we should not import components from other actors.

File: packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts [hunk 4]
```diff
@@ -217,4 +222,67 @@ describe('ActorQueryOperationLeftJoin', () => {
       });
     });
   });
+
+  describe('Comunica sparql regression test for #1029', () => {
```
Comment (rubensworks) [inline packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts:226]: This test belongs in the system tests (`engines/query-sparql`)

File: packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts (hunk 2 above)
Comment (heartpunk) [reply packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts:14]: fixed

File: packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts (hunk 3 above)
Comment (heartpunk) [reply packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts:18]: fixed.

**Comment by heartpunk** [comment]
No immediate clue as to what to do with respect to the test coverage gap. The query engine factory wasn't modified by this change AFAICT? That does seem to be the next blocking issue though, so I'll look at that more next.

File: engines/query-sparql/test/QuerySparql-test.ts [hunk 5]
```diff
@@ -9,10 +9,11 @@ import type { QueryBindings, QueryStringContext } from '@comunica/types';
 import type * as RDF from '@rdfjs/types';
 import 'jest-rdf';
 import arrayifyStream from 'arrayify-stream';
+import * as n3 from 'n3';
 import { Store } from 'n3';
 import { DataFactory } from 'rdf-data-factory';
 import { Factory } from 'sparqlalgebrajs';
-import { QueryEngine } from '../lib/QueryEngine';
+import { QueryEngine } from '../lib';
```
Comment (rubensworks) [inline engines/query-sparql/test/QuerySparql-test.ts:16]: This _may_ be the cause of the coverage drop.

File: packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts [hunk 6]
```diff
@@ -6,7 +6,7 @@ import { ArrayIterator, UnionIterator } from 'asynciterator';
 import { DataFactory } from 'rdf-data-factory';
 import * as sparqlee from 'sparqlee';
 import { isExpressionError } from 'sparqlee';
-import { ActorQueryOperationLeftJoin } from '../lib/ActorQueryOperationLeftJoin';
+import { ActorQueryOperationLeftJoin } from '../lib';
```
Comment (rubensworks) [inline packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts:9]: This change also shouldn't be needed.

Comment (heartpunk) [reply packages/actor-query-operation-leftjoin/test/ActorQueryOperationLeftJoin-test.ts:9]: gtk, webstorm suggested it, otherwise i wouldn't have done this.

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 5 above)
Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts:16]: seems to fix it!

**Comment by heartpunk** [comment]
I assume you'd like this all squashed into a single commit at this point? Any suggestion for commit message? Lacking more holistic knowledge of the codebase I'm a bit unsure about the best choice there.

**Comment by rubensworks** [comment]
> I assume you'd like this all squashed into a single commit at this point? Any suggestion for commit message? Lacking more holistic knowledge of the codebase I'm a bit unsure about the best choice there.

It's fine like this, we can easily squash via the GH UI :-)

File: engines/query-sparql/test/QuerySparql-test.ts [hunk 7]
```diff
@@ -790,6 +791,70 @@ describe('System test: QuerySparql', () => {
     });
   });
 
+  describe('Comunica sparql regression test for #1029', () => {
```
Comment (rubensworks) [inline engines/query-sparql/test/QuerySparql-test.ts:794]: The issue reference might be a bit confusing, as it's not immediately clear as to what the issue is.
Can we move this test into the `query` describe block?

File: engines/query-sparql/test/QuerySparql-test.ts [hunk 8]
```diff
@@ -790,6 +791,70 @@ describe('System test: QuerySparql', () => {
...
+          FILTER EXISTS {
+            ?s2 ?p ?o.
+            FILTER (?s != ?s2)
+          }
+        }
+      }
+      `,
+      { sources: [ store ]});
+      const result_quads = await quadStream.toArray();
+      expect(result_quads).toEqual([{
```
Comment (rubensworks) [inline engines/query-sparql/test/QuerySparql-test.ts:821]: Could the quads and terms be constructed using the data factory (like in the other tests?)

File: packages/actor-query-operation-leftjoin/package.json [hunk 9]
```diff
@@ -31,10 +31,10 @@
     "lib/**/*.js"
   ],
   "dependencies": {
-    "@comunica/bus-query-operation": "^2.4.0",
-    "@comunica/bus-rdf-join": "^2.4.0",
-    "@comunica/core": "^2.4.0",
-    "@comunica/types": "^2.4.0",
+    "@comunica/bus-query-operation": "^2.3.0",
```
Comment (rubensworks) [inline packages/actor-query-operation-leftjoin/package.json:34]: I assume this is because your branch is out-of-sync with master?

**Review by rubensworks — CHANGES_REQUESTED** [review]
Just some minor things. Other than that, this should be good to go!
Nice work @heartpunk!

File: packages/actor-query-operation-leftjoin/package.json (hunk 9 above)
Comment (rubensworks) [reply packages/actor-query-operation-leftjoin/package.json:34]: Related to the yarn.lock merge conflict mentioned below.

Comment (heartpunk) [reply packages/actor-query-operation-leftjoin/package.json]: yep, fixed this.

**Comment by heartpunk** [comment]
dealt with the branch drift, gonna finish up the rest later today!

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 7 above)
Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts]: i've moved it, but the description prolly needs some updating still. not resolving this thread until i finish that.

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 8 above)
Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts]: absolutely, next step for me today!

Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts]: close!!

**Comment by heartpunk** [comment]
currently has a type error, but much closer to conventions, and a useful stopping point of reference for my workflow

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 8 above)
Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts]: should be addressed

**Comment by heartpunk** [comment]
the test failure is unclear to me; seems related to some changes i rebased over and perhaps merged improperly at a point? thoughts @rubensworks?

**Comment by heartpunk** [comment]
ok i tried to clean the history up a little, but the test failure remains. i await ur comment before further changes at this point!

**Comment by rubensworks** [comment]
You can ignore the test failure, it's unrelated to your PR: https://github.com/comunica/comunica/issues/1106

**Comment by heartpunk** [comment]
Cool, I think it's probably succeeding on all other tests aside from that one, so we can review/squash/merge at your leisure. I'll try to rebase it periodically until then.

File: engines/query-sparql/test/QuerySparql-test.ts [hunk 10]
```diff
@@ -573,6 +573,43 @@ describe('System test: QuerySparql', () => {
         expect((await arrayifyStream(await result.execute()))).toEqual([]);
       });
     });
+
+    // TODO: better name here before merge.
```
Comment (rubensworks) [inline engines/query-sparql/test/QuerySparql-test.ts:577]: TODO

File: engines/query-sparql/test/QuerySparql-test.ts [hunk 11]
```diff
@@ -573,6 +573,43 @@ describe('System test: QuerySparql', () => {
...
+        const nodes = [ DF.quad(s, p, o), DF.quad(b, p, o) ];
+
+        const store = new Store(nodes);
+
+        const result = await engine.queryQuads(
+          `CONSTRUCT {?s ?p ?o.} WHERE {?s ?p ?o. OPTIONAL {FILTER EXISTS {?s2 ?p ?o. FILTER (?s != ?s2)}}}`,
+          { sources: [ store ]},
+        );
+        expect(await arrayifyStream(result)).toEqual([
+          {
```
Comment (rubensworks) [inline engines/query-sparql/test/QuerySparql-test.ts:594]: These should be made with a DataFactory.

File: package.json [hunk 12]
```diff
@@ -8,12 +8,9 @@
   "devDependencies": {
     "@babel/core": "^7.16.0",
     "@babel/preset-env": "^7.16.4",
-    "@inrupt/solid-client-authn-core": "^1.12.2",
```
Comment (rubensworks) [inline package.json:11]: Incorrect removals?

File: yarn.lock [hunk 13]
```diff
@@ -994,43 +994,6 @@
   resolved "https://registry.yarnpkg.com/@hutson/parse-repository-url/-/parse-repository-url-3.0.2.tgz#98c23c950a3d9b6c8f0daed06da6c3af06981340"
   integrity sha512-H9XAx3hc0BQHY6l+IFSWHDySypcXsvsuLhgYLUGywmJ5pswRVQJUHpOsobnLYp2ZUaUlKiKDrgWWhosOwAEM8Q==
 
-"@inrupt/solid-client-authn-core@^1.12.2":
```
Comment (rubensworks) [inline yarn.lock:997]: These removals are also not necessary.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Thanks! Almost there.

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 11 above)
Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts:594]: i was unable to get the tests to pass that way, but i'll give that more work! the blank nodes ended up with hard to predict literal names and that caused the equality tests to fail.

File: package.json (hunk 12 above)
Comment (heartpunk) [reply package.json:11]: this was done by a yarn run (idk which command unfort)! i was surprised that yarn did it and didn't expect it to be needed. i'll revert it back to how it was before.

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 11 above)
Comment (rubensworks) [reply engines/query-sparql/test/QuerySparql-test.ts:594]: We probably don't need the blank nodes? All named nodes should be fine.

Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts:594]: mk, switching it over.

Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts:594]: gonna push a failing version up so you can point out the fix; it's obviously a fairly trivial cause of failure given the size of the diff. my suspicion is that the sparql construct query in the test (also, full disclosure: i don't really know SPARQL!) and the data factory for named nodes just don't operate in the same way.

tl;dr: our fixture makes named nodes where the sparql query seems to make objects (??? not sure what that means in this context, but, it is the word the diff uses)

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 10 above)
Comment (heartpunk) [reply engines/query-sparql/test/QuerySparql-test.ts:577]: indeed. given my low SPARQL familiarity, i do not have a great idea. i'm enthusiastic about helping, and also i know my limits. could you suggest something please? it would actually help me understand the nature of this issue better too, and i'd love to keep contributing more in the future.

Comment (rubensworks) [reply engines/query-sparql/test/QuerySparql-test.ts:577]: Something like `'filter expression inside an optional'` should be fine.

File: engines/query-sparql/test/QuerySparql-test.ts [hunk 14]
```diff
@@ -573,6 +573,29 @@ describe('System test: QuerySparql', () => {
...
+
+        const nodes = [ DF.quad(s, p, o), DF.quad(s1, p, o) ];
+
+        const store = new Store(nodes);
+
+        const result = await engine.queryQuads(
+          `CONSTRUCT {?s ?p ?o.} WHERE {?s ?p ?o. OPTIONAL {FILTER EXISTS {?s2 ?p ?o. FILTER (?s != ?s2)}}}`,
+          { sources: [ store ]},
+        );
+        expect(await arrayifyStream(result)).toEqual([
```
Comment (rubensworks) [inline engines/query-sparql/test/QuerySparql-test.ts:593]:
```suggestion
        expect(await arrayifyStream(result)).toEqualRdfQuadArray([
```

Changing the equality checker should resolve the problem.

More details: https://github.com/rubensworks/jest-rdf

File: engines/query-sparql/test/QuerySparql-test.ts (hunk 11 above)
Comment (rubensworks) [reply engines/query-sparql/test/QuerySparql-test.ts]: See https://github.com/comunica/comunica/pull/1046/files#r1067792166

**Comment by heartpunk** [comment]
ok, lint fixes and a new test failure to track down. will poke at it today!

**Comment by heartpunk** [comment]
`yarn install` is removing inrupt/solid dependencies again. Do we really want those removed? should yarn be modifying dependencies like that? @rubensworks

**Comment by heartpunk** [comment]
(I'm having to rerun it because of environment changes since I last worked on this.)

**Comment by rubensworks** [comment]
> yarn install is removing inrupt/solid dependencies again. Do we really want those removed? should yarn be modifying dependencies like that? @rubensworks

They shouldn't actually be removed, but it's because we can't install them on Node 18.
The hack is not ideal, but I'm open to alternatives! :-)

**Comment by heartpunk** [comment]
dropping back to 17 to do the fixes!

**Comment by heartpunk** [comment]
ah, guess there's an even odd scheme bcz 17 gives a different warning, going to 16 then.

**Comment by heartpunk** [comment]
I think I've messed up the branch a lil by way of git rustiness, working on the fixes now that I'm a bit more up to speed on the context of the pull request again.

**Comment by rubensworks** [comment]
Closed because of https://github.com/comunica/comunica/issues/1029#issuecomment-1460433624
